lib/v2/douyin/utils.js

```diff
@@ -42,9 +46,88 @@ const getOriginAvatar = (url) =>
         .replace(/^(.*\.douyinpic\.com\/).*(\/aweme-avatar\/)([^?]*)(\?.*)?$/, '$1origin$2$3')
         .replace(/~\w+_\d+x\d+/g, '');
 
+const extractRenderData = (html) => {
+    if (!html) {
+        throw Error('Empty html. The request may be filtered by WAF.');
+    }
+
+    const $ = cheerio.load(html);
+    const renderData = JSON.parse(decodeURIComponent($('script#RENDER_DATA').html()));
+    if (!renderData) {
+        const title = $('title').text();
+        if (title.includes('验证')) {
+            throw Error(title);
+        } else if ($('#captcha_container').length > 0) {
+            throw Error('Captcha required.');
+        }
+        throw Error('Failed to get render data.');
+    } else if (renderData.isSpider) {
+        throw Error('The request was considered to be from a spider.');
+    } else {
+        return renderData;
+    }
+};
+
+const gotGet = async (url) => {
+    const response = await got(url, {
+        headers: {
+            'User-Agent': 'curl/7.86.0', // magic!
+        },
+    });
+    return extractRenderData(response.data);
+};
+
+const puppeteerGet = async (pageUrl) => {
+    const browser = await require('@/utils/puppeteer')();
+    const page = await browser.newPage();
+    await page.setRequestInterception(true);
+    page.on('request', (request) => {
+        request.resourceType() === 'document' ? request.continue() : request.abort();
+    });
+    await page.goto(pageUrl, {
+        waitUntil: 'domcontentloaded',
+    });
+    const html = await page.evaluate(() => document.documentElement.innerHTML);
+    await browser.close();
+    return extractRenderData(html);
+};
+
+const universalGet = async (url, route, ctx) => {
+    const cacheKey = `douyin:utils:universalGet:gotFirst:${route}`;
+    const gotFirst = await ctx.cache.tryGet(
+        cacheKey,
+        () => (route === 'live' ? '0' : '1'), // 'live' route has no magic
```

Comment:
For 'live', looks like their WAF will let go if UA is Googlebot `Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)` and `#RENDER_DATA` is present. However, that's not the case for 'user' and 'hashtag'.